File: python/lightning/client/_validation.py
from __future__ import annotations

import json
from typing import Any, Optional


class ValidationError(ValueError):
    pass


MAX_ID_LENGTH = 1024
MAX_CONTENT_LENGTH = 10 * 1024 * 1024  # 10MB
MAX_ENTITY_TYPE_LENGTH = 256
MAX_METADATA_SIZE = 1024 * 1024  # 1MB
MAX_EMBEDDING_DIM = 16384
MAX_QUERY_LENGTH = 100_000


def validate_id(id: str, field_name: str = "id") -> None:
    if not isinstance(id, str):
        raise ValidationError(f"{field_name} must be a string, got {type(id).__name__}")
    if not id:
        raise ValidationError(f"{field_name} must not be empty")
    if len(id) > MAX_ID_LENGTH:
        raise ValidationError(f"{field_name} exceeds max length of {MAX_ID_LENGTH}")


def validate_content(content: str, field_name: str = "content") -> None:
    if not isinstance(content, str):
        raise ValidationError(f"{field_name} must be a string, got {type(content).__name__}")
    if len(content) > MAX_CONTENT_LENGTH:
        raise ValidationError(f"{field_name} exceeds max {MAX_CONTENT_LENGTH} bytes")


def validate_entity_type(entity_type: str) -> None:
    if not isinstance(entity_type, str):
        raise ValidationError(f"entity_type must be a string, got {type(entity_type).__name__}")
    if not entity_type:
        raise ValidationError("entity_type must not be empty")
    if len(entity_type) > MAX_ENTITY_TYPE_LENGTH:
        raise ValidationError(f"entity_type exceeds max length of {MAX_ENTITY_TYPE_LENGTH}")


def validate_metadata(metadata: Any) -> str:
    if isinstance(metadata, str):
        if len(metadata) > MAX_METADATA_SIZE:
            raise ValidationError(f"metadata exceeds max {MAX_METADATA_SIZE} bytes")
        try:
            json.loads(metadata)
            return metadata
        except json.JSONDecodeError as e:
            raise ValidationError(f"metadata is not valid JSON: {e}")
    serialized = json.dumps(metadata, default=str)
    if len(serialized) > MAX_METADATA_SIZE:
        raise ValidationError(f"metadata exceeds max {MAX_METADATA_SIZE} bytes after serialization")
    return serialized


def validate_embedding(embedding: Optional[list[float]]) -> None:
    if embedding is None:
        return
    if not isinstance(embedding, list):
        raise ValidationError(f"embedding must be a list of floats, got {type(embedding).__name__}")
    if not embedding:
        raise ValidationError("embedding must not be empty if provided")
    if len(embedding) > MAX_EMBEDDING_DIM:
        raise ValidationError(f"embedding dimension {len(embedding)} exceeds max {MAX_EMBEDDING_DIM}")
    for i, v in enumerate(embedding):
        if not isinstance(v, (int, float)):
            raise ValidationError(f"embedding[{i}] must be numeric, got {type(v).__name__}")


def validate_top_k(top_k: int, max_allowed: int = 1000) -> None:
    if not isinstance(top_k, int):
        raise ValidationError(f"top_k must be an integer, got {type(top_k).__name__}")
    if top_k < 1:
        raise ValidationError("top_k must be >= 1")
    if top_k > max_allowed:
        raise ValidationError(f"top_k {top_k} exceeds max allowed {max_allowed}")


def validate_query_string(query: str) -> None:
    if not isinstance(query, str):
        raise ValidationError(f"query must be a string, got {type(query).__name__}")
    if len(query) > MAX_QUERY_LENGTH:
        raise ValidationError(f"query exceeds max length of {MAX_QUERY_LENGTH}")


def validate_batch_entities(entities: list[dict], max_batch: int = 1000) -> None:
    if not isinstance(entities, list):
        raise ValidationError(f"entities must be a list, got {type(entities).__name__}")
    if not entities:
        raise ValidationError("entities list must not be empty")
    if len(entities) > max_batch:
        raise ValidationError(f"batch size {len(entities)} exceeds max {max_batch}")
    for i, e in enumerate(entities):
        if not isinstance(e, dict):
            raise ValidationError(f"entities[{i}] must be a dict, got {type(e).__name__}")
        if "id" not in e:
            raise ValidationError(f"entities[{i}] is missing required 'id' field")
        if "content" not in e:
            raise ValidationError(f"entities[{i}] is missing required 'content' field")
        validate_id(e["id"], f"entities[{i}].id")
        validate_content(e["content"], f"entities[{i}].content")


def validate_hops(hops: int) -> None:
    if not isinstance(hops, int):
        raise ValidationError(f"hops must be an integer, got {type(hops).__name__}")
    if hops < 1:
        raise ValidationError("hops must be >= 1")
    if hops > 10:
        raise ValidationError("hops must not exceed 10 (exponential traversal guard)")


def validate_store_params(
    id: str,
    content: str,
    entity_type: str,
    metadata: Any,
    embedding: Optional[list[float]],
) -> None:
    validate_id(id)
    validate_content(content)
    validate_entity_type(entity_type)
    validate_metadata(metadata)
    validate_embedding(embedding)

File: packages/lightning-client/src/validation.ts
export class ValidationError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'ValidationError'
  }
}

const MAX_ID_LENGTH = 1024
const MAX_CONTENT_LENGTH = 10 * 1024 * 1024
const MAX_ENTITY_TYPE_LENGTH = 256
const MAX_EMBEDDING_DIM = 16384
const MAX_QUERY_LENGTH = 100_000

export function validateId(id: unknown, fieldName = 'id'): asserts id is string {
  if (typeof id !== 'string') {
    throw new ValidationError(`${fieldName} must be a string, got ${typeof id}`)
  }
  if (id.length === 0) {
    throw new ValidationError(`${fieldName} must not be empty`)
  }
  if (id.length > MAX_ID_LENGTH) {
    throw new ValidationError(`${fieldName} exceeds max length of ${MAX_ID_LENGTH}`)
  }
}

export function validateContent(content: unknown, fieldName = 'content'): asserts content is string {
  if (typeof content !== 'string') {
    throw new ValidationError(`${fieldName} must be a string, got ${typeof content}`)
  }
  if (content.length > MAX_CONTENT_LENGTH) {
    throw new ValidationError(`${fieldName} exceeds max ${MAX_CONTENT_LENGTH} bytes`)
  }
}

export function validateEntityType(entityType: unknown): asserts entityType is string {
  if (typeof entityType !== 'string') {
    throw new ValidationError(`entity_type must be a string, got ${typeof entityType}`)
  }
  if (entityType.length === 0) {
    throw new ValidationError('entity_type must not be empty')
  }
  if (entityType.length > MAX_ENTITY_TYPE_LENGTH) {
    throw new ValidationError(`entity_type exceeds max length of ${MAX_ENTITY_TYPE_LENGTH}`)
  }
}

export function validateEmbedding(embedding: unknown): embedding is number[] | undefined {
  if (embedding === undefined || embedding === null) return true
  if (!Array.isArray(embedding)) {
    throw new ValidationError(`embedding must be an array of floats, got ${typeof embedding}`)
  }
  if (embedding.length === 0) {
    throw new ValidationError('embedding must not be empty if provided')
  }
  if (embedding.length > MAX_EMBEDDING_DIM) {
    throw new ValidationError(`embedding dimension ${embedding.length} exceeds max ${MAX_EMBEDDING_DIM}`)
  }
  for (let i = 0; i < embedding.length; i++) {
    if (typeof embedding[i] !== 'number' || !Number.isFinite(embedding[i])) {
      throw new ValidationError(`embedding[${i}] must be a finite number`)
    }
  }
  return true
}

export function validateTopK(topK: unknown, maxAllowed = 1000): asserts topK is number {
  if (typeof topK !== 'number' || !Number.isInteger(topK)) {
    throw new ValidationError(`top_k must be an integer, got ${typeof topK}`)
  }
  if (topK < 1) {
    throw new ValidationError('top_k must be >= 1')
  }
  if (topK > maxAllowed) {
    throw new ValidationError(`top_k ${topK} exceeds max allowed ${maxAllowed}`)
  }
}

export function validateQueryString(query: unknown): asserts query is string {
  if (typeof query !== 'string') {
    throw new ValidationError(`query must be a string, got ${typeof query}`)
  }
  if (query.length > MAX_QUERY_LENGTH) {
    throw new ValidationError(`query exceeds max length of ${MAX_QUERY_LENGTH}`)
  }
}

export function validateHops(hops: unknown): asserts hops is number {
  if (typeof hops !== 'number' || !Number.isInteger(hops)) {
    throw new ValidationError(`hops must be an integer, got ${typeof hops}`)
  }
  if (hops < 1) {
    throw new ValidationError('hops must be >= 1')
  }
  if (hops > 10) {
    throw new ValidationError('hops must not exceed 10 (exponential traversal guard)')
  }
}

export function validateBatchEntities(entities: unknown, maxBatch = 1000): asserts entities is Record<string, unknown>[] {
  if (!Array.isArray(entities)) {
    throw new ValidationError(`entities must be an array, got ${typeof entities}`)
  }
  if (entities.length === 0) {
    throw new ValidationError('entities list must not be empty')
  }
  if (entities.length > maxBatch) {
    throw new ValidationError(`batch size ${entities.length} exceeds max ${maxBatch}`)
  }
  for (let i = 0; i < entities.length; i++) {
    const e = entities[i]
    if (typeof e !== 'object' || e === null) {
      throw new ValidationError(`entities[${i}] must be an object`)
    }
    if (!('id' in e) || typeof (e as Record<string, unknown>).id !== 'string') {
      throw new ValidationError(`entities[${i}] is missing required 'id' field`)
    }
    if (!('content' in e) || typeof (e as Record<string, unknown>).content !== 'string') {
      throw new ValidationError(`entities[${i}] is missing required 'content' field`)
    }
  }
}

export function validateMetadata(metadata: unknown): string {
  if (typeof metadata === 'string') {
    try {
      JSON.parse(metadata)
      return metadata
    } catch {
      throw new ValidationError('metadata is not valid JSON')
    }
  }
  return JSON.stringify(metadata ?? {})
}
